File: lib/java/src/main/java/io/github/msgpact/internal/schema/ParseField.java
package io.github.msgpact.internal.schema;

import static io.github.msgpact.internal.schema.GetTypeUnexpectedParseFailure.getTypeUnexpectedParseFailure;
import static io.github.msgpact.internal.schema.ParseTypeDeclaration.parseTypeDeclaration;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

import io.github.msgpact.MsgPactSchemaParseError;
import io.github.msgpact.internal.types.VFieldDeclaration;
import io.github.msgpact.internal.types.VType;

public class ParseField {
    static VFieldDeclaration parseField(
            List<Object> path,
            String fieldDeclaration,
            Object typeDeclarationValue,
            ParseContext ctx) {
        final var regexString = "^([a-z][a-zA-Z0-9_]*)(!)?$";
        final var regex = Pattern.compile(regexString);

        final var matcher = regex.matcher(fieldDeclaration);
        if (!matcher.find()) {
            final List<Object> finalPath = new ArrayList<>(path);
            finalPath.add(fieldDeclaration);

            throw new MsgPactSchemaParseError(List.of(new SchemaParseFailure(ctx.documentName, finalPath,
                    "KeyRegexMatchFailed", Map.of("regex", regexString))), ctx.msgPactSchemaDocumentNamesToJson);
        }

        final var fieldName = matcher.group(0);
        final var optional = matcher.group(2) != null;

        final List<Object> thisPath = new ArrayList<>(path);
        thisPath.add(fieldName);

        if (!(typeDeclarationValue instanceof List)) {
            throw new MsgPactSchemaParseError(
                    getTypeUnexpectedParseFailure(ctx.documentName, thisPath, typeDeclarationValue, "Array"),
                    ctx.msgPactSchemaDocumentNamesToJson);
        }
        final List<Object> typeDeclarationArray = (List<Object>) typeDeclarationValue;

        final var typeDeclaration = parseTypeDeclaration(
                thisPath,
                typeDeclarationArray,
                ctx);

        return new VFieldDeclaration(fieldName, typeDeclaration, optional);
    }
}

File: lib/py/telepact/internal/schema/ParseField.py
import re
from typing import TYPE_CHECKING
from ...internal.schema.SchemaParseFailure import SchemaParseFailure
from ..types.VFieldDeclaration import VFieldDeclaration

if TYPE_CHECKING:
    from ..types.VType import VType
    from ...internal.schema.ParseContext import ParseContext


def parse_field(path: list[object], field_declaration: str, type_declaration_value: object,
                ctx: 'ParseContext') -> 'VFieldDeclaration':
    from ...TelepactSchemaParseError import TelepactSchemaParseError
    from ...internal.schema.GetTypeUnexpectedParseFailure import get_type_unexpected_parse_failure
    from ...internal.schema.ParseTypeDeclaration import parse_type_declaration

    regex_string = r"^([a-z][a-zA-Z0-9_]*)(!)?$"
    regex = re.compile(regex_string)

    matcher = regex.match(field_declaration)
    if not matcher:
        final_path = path + [field_declaration]
        raise TelepactSchemaParseError([SchemaParseFailure(ctx.document_name, final_path,
                                                          "KeyRegexMatchFailed",
                                                          {"regex": regex_string})], ctx.telepact_schema_document_names_to_json)

    field_name = matcher.group(0)
    optional = bool(matcher.group(2))

    this_path = path + [field_name]

    if not isinstance(type_declaration_value, list):
        raise TelepactSchemaParseError(get_type_unexpected_parse_failure(
            ctx.document_name,
            this_path,
            type_declaration_value,
            "Array"), ctx.telepact_schema_document_names_to_json)
    type_declaration_array = type_declaration_value

    type_declaration = parse_type_declaration(this_path,
                                              type_declaration_array, ctx)

    return VFieldDeclaration(field_name, type_declaration, optional)
